### DISCUSSION TITLE
Allow to attach deadlines to questions

### DISCUSSION THREAD

**Discussion by rixx** [pretalx/pretalx#1069] [Ideas] [answered]
This is an issue that is well-suited for external contributors – let me know if I can be of any help, either via this issue or via support@pretalx.com.

## What this is about

pretalx provides a [Question](https://docs.pretalx.org/en/latest/developer/interfaces/models.html#questions-and-answers) model to get arbitrary information from speakers or reviewers.

A deadline would make nagging emails potentially easier. It could also make an answer optional before that deadline, and mandatory (form-wise) afterwards. Things like travel details make mostly no sense to ask early on, but might be required facts close to the event.

## How to get started

First of all, leave a comment on this issue to make sure nobody else starts working on this issue at the same time!

Next, you need pretalx running locally on your machine. First, [fork](https://github.com/pretalx/pretalx/fork) the repository to your own GitHub account. Then clone it locally (or [set it as your remote](https://help.github.com/articles/configuring-a-remote-for-a-fork/) if you already cloned it earlier)

Follow the [developer setup guide](https://docs.pretalx.org/en/latest/developer/setup.html), and then play around with pretalx a bit.
Have a look at the places impacting this issue. Look at the code responsible for the issues (searching for relevant words or files should help you find your way around). We have an explanation of our [code structure](https://docs.pretalx.org/en/latest/developer/concepts.html) and the actual [data models](https://docs.pretalx.org/en/latest/developer/interfaces/models.html).

## How to do this

The Question model is in `pretalx/submission/models/question.py`, and the `deadline` field would be a nullable `DateTimeField`. It should be clear that leaving the deadline empty will set no deadline at all.

There are two things we might expect from a question with a deadline:

 -   The answer could be optional before the deadline and mandatory after the deadline
 -   The answer could be allowed before the deadline and frozen after the deadline

The second behaviour is probably the expected one from the word "deadline", but at the same time disallowing answers will just produce support requests from the speakers who are too late, so there should also be a `deadline_mode` field detailing this.

Or, to be more clear, we could add two different fields: `require_after` and `freeze_after`, both optional `DateTimeField`s.

You'll then need to look at the `QuestionForm` and make sure the read_only and required attributes are set correctly on the fields generated by it.

## Submit the PR

There are a number of tutorials on the Pull Request workflow online, for example this [short version](https://guides.github.com/introduction/flow/) by GitHub itself and this [more in-depth tutorial](https://www.digitalocean.com/community/tutorials/how-to-create-a-pull-request-on-github).

Remember to provide documentation and tests, if applicable. Run the tests before opening your PR, including style checks. We have a [contribution guide](https://docs.pretalx.org/en/latest/developer/contributing.html), which ought to prove helpful at this stage.

**Comment by nataliaKat** [comment]
Hello, we would like to give this issue a try together with @AthinaDavari.

**Reply by rixx** [reply]
Cool! Let me know if you have any questions.

**Reply by AthinaDavari** [reply]
Hi,
we made two fields in questions' model (and its form) for the change of the feature
* <code>question_required</code>, a charfield
* <code>deadline</code>, a datetime field

![imageee](https://user-images.githubusercontent.com/56487775/117539431-47c6a500-b013-11eb-94ae-fa5c1494dac7.png)

And with those changes we will delete the <code>required</code> boolean field (checkbox in form).

Now we want to apply the changes in answer forms (required or optional or freeze). Can you give us advice how to continue on contributing?

If you want to see our changes, we can either make a draft PR or you can visit [our fork](https://github.com/AthinaDavari/pretalx/tree/Attach_Deadlines_Questions).

Thanks in advance

**Reply by nataliaKat** [reply]
Hello again @rixx , 
We have made some progress since the last reply. 
- We removed the <code>required</code> field, since all cases are covered with <code>question_required</code>. 
- Also we modified <code>src/pretalx/common/mixins/forms.py</code> so that the question is required and disabled based on the <code>question_required</code> field and the current date compared to the <code>deadline</code>. 
- Lastly, we fixed some tests which failed due to our changes.

We would really appreciate it if you gave us some feedback related to our work and some advice on how to test the new functionality with the deadlines. We would be glad if we finished with the changes till Friday, in order to include them in our final presentation for our Software Engineering in Practice course, which is on Monday 28/06. We can either make a draft pull request or you can visit [our fork](https://github.com/AthinaDavari/pretalx/tree/Attach_Deadlines_Questions).

Thanks in advance!

**Reply by rixx** [reply]
Ni @nataliaKat – I completely missed your last update, sorry for that! I'll definitely review your PR this week, and I'm confident that we'll find a way to merge it before your presentation :)

**Reply by nataliaKat** [reply]
Thank you very much!!! We'll make a draft PR so that you can see our changes and we can change anything needed. Also we would like some advice related to tests.

**Comment by rixx** [comment, marked as answer]
Implemented and will be part of pretalx 2.2.0.
